### PR TITLE
[skip uplift] #35302: Add LLK support for lossless fp32 tilize

### PR DESCRIPTION
**Pull request by pmilenkovicTT**
### Ticket
[#35302](https://github.com/tenstorrent/tt-metal/issues/35302) (tt-metal issue)
### Problem description

Float32 datums were getting converted to tf32/float16 during the tilize operation, resulting in accuracy loss.

### What's changed
Enabled lossless tilize for Float32 by following the same unpack-to-dest path used for Int32/UInt32, but triggered when `unpack_dst_format == Float32`. 
This also requires the `_llk_unpack_tilize_` function to take additional parameter, leading to trivial changes in test infra. 

LLK changes (Wormhole B0 & Blackhole):

- `_llk_unpack_tilize_init_`: Added Float32 check to unpack_to_dest condition
- `_llk_unpack_tilize_`: Added `unpack_dst_format` parameter to enable runtime Float32 detection

Test infra updates:

- Updated all `_llk_unpack_tilize_` call sites to pass the new `unpack_dst_format` parameter

This requires corresponding changes in tt-metal:

- LLK API layer (`llk_unpack_tilize_api.h`) to pass `unpack_dst_format`
- Program factories to set proper `unpack_dst_format` in CBs

This is tested for WH. Making this a draft PR while testing for BH. 
Edit: Works on both WH & BH. 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist

- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)

### CODE REVIEW

File: tt_llk_wormhole_b0/llk_lib/llk_unpack_tilize.h [hunk 1]
```diff
@@ -182,15 +184,18 @@ inline void _llk_unpack_tilize_(
...
     const bool narrow_tile          = false)
 {
     LLK_ASSERT(num_faces == 1 || num_faces == 2 || num_faces == 4, "num_faces must be 1, 2, or 4");
-    // In case of 32-bit integer numbers, we have to unpack into dest register
+    // In case of 32-bit numbers, we have to unpack into dest register
+    // For integers, always unpack to dest. For Float32, only if unpack_dst_format is Float32 (lossless tilize mode)
     const bool unpack_to_dest = (unpack_src_format == static_cast<std::underlying_type_t<DataFormat>>(DataFormat::UInt32)) ||
-                                (unpack_src_format == static_cast<std::underlying_type_t<DataFormat>>(DataFormat::Int32));
+                                (unpack_src_format == static_cast<std::underlying_type_t<DataFormat>>(DataFormat::Int32)) ||
+                                (unpack_dst_format == static_cast<std::underlying_type_t<DataFormat>>(DataFormat::Float32));
```
[the remark this answers is not in the document]
Comment (pmilenkovicTT) [reply tt_llk_wormhole_b0/llk_lib/llk_unpack_tilize.h]: @rtawfik01 Any idea if Copilot makes a good point here? I don't think we can actually get into the situation where we want to tilize non-float32 source but make dest float32.

File: tt_llk_blackhole/llk_lib/llk_unpack_tilize.h [hunk 2]
```diff
@@ -50,9 +50,11 @@ inline void _llk_unpack_tilize_init_(
 
     const std::uint32_t block_c_dim = ct_dim * (narrow_tile ? FACE_C_DIM : TILE_C_DIM);
 
-    // In case of 32-bit integer numbers, we have to unpack into dest register
+    // In case of 32-bit numbers, we have to unpack into dest register
+    // For integers, always unpack to dest. For Float32, only if unpack_dst_format is Float32 (lossless tilize mode)
     const bool unpack_to_dest = (unpack_src_format == static_cast<std::underlying_type_t<DataFormat>>(DataFormat::UInt32)) ||
```
[the remark this answers is not in the document]
Comment (ncvetkovicTT) [reply tt_llk_blackhole/llk_lib/llk_unpack_tilize.h]: Thanks! If you have any other suggestion that will better suit kernel devs, pls add your thoughts

File: tests/python_tests/test_unpack_tilize.py [hunk 3]
```diff
@@ -67,17 +87,29 @@ def unpack_tilize(formats, workers_tensix_coordinates):
...
-    ), "Result tensor and golder tensor are not of the same length"
+    ), "Result tensor and golden tensor are not of the same length"
 
     res_tensor = torch.tensor(res_from_L1, dtype=format_dict[formats.output_format])
 
-    assert passed_test(
-        golden_tensor, res_tensor, formats.output_format
-    ), "Assert against golden failed"
+    if validate_lossless:
+        # Lossless validation
```
Comment (fvranicTT) [inline tests/python_tests/test_unpack_tilize.py]:
```python
diff = golden_tensor - res_tensor
abs_diff = diff.abs()
assert torch.allclose(golden_tensor, res_tensor, atol=0, rtol=1e-6), (
    f"Float32 tilize lost precision! Input and output differ.\n"
    f"Max difference: {abs_diff.max().item()}\n"
    f"Num different elements: {(abs_diff > 1e-6).sum()}\n"
    f"Expected (golden): {golden_tensor[:10]}\n"
    f"Got (result): {res_tensor[:10]}"
)
```
No need to compute the diff twice.
